src/com/google/javascript/jscomp/AdvancedCoverageInstrumentationCallback.java

```diff
@@ -0,0 +1,205 @@
+    // can not call the Code instrumentation function on itself
+    if (sourceFileName.endsWith("base.js") || sourceFileName.endsWith(INSTRUMENT_CODE_FILE_NAME)) {
+      return;
+    }
+
+    if (node.isScript()) {
+      if (instrumentationData.get(fileName) != null) {
+        compiler.reportChangeToEnclosingScope(node);
+        instrumentCode(traversal, instrumentationData.get(fileName));
+      }
+    }
+
+    if (node.isFunction()) {
+
+      functionName = NodeUtil.getBestLValueName(NodeUtil.getBestLValue(node));
+
+      if (!instrumentationData.containsKey(fileName)) {
+        instrumentationData.put(fileName, new AdvancedFileInstrumentationData(fileName));
+      }
+      addBlockNode(instrumentationData.get(fileName), Arrays.asList(node.getLastChild()),
+          functionName);
+
+    }
+  }
+
+  /**
+   * Iterate over all collected block nodes within a Script node and add a new child to the front of
+   * each block node which is the instrumentation Node
+   *
+   * @param traversal Tne node traversal context which maintains information such as fileName being
+   *                  traversed
+   * @param data      Data structure that maintains an organized list of block nodes that need to be
+   *                  instrumented
+   */
+  private void instrumentCode(NodeTraversal traversal, AdvancedFileInstrumentationData data) {
+    Map<String, List<Node>> functionScopedNodes = data.getFunctionScopeNodes();
+    for (String key : functionScopedNodes.keySet()) {
+      for (Node block : functionScopedNodes.get(key)) {
+        block.addChildToFront(
+            newInstrumentationNode(traversal, block, key));
+        compiler.reportChangeToEnclosingScope(block);
+      }
+    }
+  }
+
+  /**
+   * Create a function call to the Instrument Code function with properly encoded parameters.
+   *
+   * @param traversal The context of the current traversal.
+   * @param node      The block node to be instrumented.
+   * @param fnName    The function name that the node exists within.
+   * @return The newly constructed function call node.
+   */
+  private Node newInstrumentationNode(NodeTraversal traversal, Node node, String fnName) {
+
+    String type = "Type.FUNCTION";
+    String combinedParam = traversal.getSourceName() + " " + fnName + " " + type;
+
+    BigInteger uniqueIdentifier = parameterMapping.getUniqueIdentifier(combinedParam);
+    BigInteger maxInteger = new BigInteger(Integer.toString(Integer.MAX_VALUE));
+
+    if (uniqueIdentifier.compareTo(maxInteger) > 0) {
+      throw new ArithmeticException(
+          "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64 VLQ");
+    }
+
+    StringBuilder sb = new StringBuilder();
+
+    try {
+      Base64VLQ.encode(sb, uniqueIdentifier.intValue());
+    } catch (IOException e) {
+      // If encoding in Base64 VLQ fails, we will use the original identifier as it will still
```

Comment:
In what cases does this fail?

This might be a good candidate for a unit test.

Reply:
Base64VLQ can potentially raise an exception because the Java Appendable `append` function can throw an `IOException`. I considered writing a unit test case, but I couldn't figure out in what way does the Java append function raise an IO exception.

Reply:
Looks like Base64VLQ shouldn't throw an exception here, because StringBuilder doesn't report that it throws an `IOException`. See also https://stackoverflow.com/questions/15607060/java-interface-throws-an-exception-but-interface-implementation-does-not-throw-a: StringBuilder doesn't have to throw an IOException even though the Appendable interface does.

My suggestion would be to catch the IOException and then throw an AssertionError in the catch block. This is the suggested "impossible condition check" from https://github.com/google/guava/wiki/ConditionalFailuresExplained.

Reply:
Interesting. I updated to throw AssertionError with message saying that an unexpected error was encounter.